juju/juju
# [JUJU-4145] Refactor the unit agent to use interfaces instead of concrete uniter api structs

The unit agent relies heavily on concrete types from the `uniter/api` package. These Unit and Uniter and Application et al structs have an embedded facade client and are not amenable to mocked tests. We want to remove JujuConnSuite so need to fix this implementation aspect first.

This PR introduces a common set of interfaces for the domain related entities - Unit, Application, Relation, RelationUnit. The api client also gets an interface. A set of shims is added to adapt the uniter api structs to the interfaces. The mostly duplicate interfaces from the subpackages are replaced by the common domain ones or else things won't compile (and the various extra shims are removed).

The end result is a unit agent that will be easier to replace JujuConnSuite in a subsequent PR.

Most of the additional LOC are the new mocks. A large chick of the churn is also renaming of various receivers. And duplicate mocks were removed.

## QA steps

run unit tests

also bootstrap and deploy and relate a couple of charms
run the leader related hook commands, eg `is-leader`, 'leader-get` etc
update relation content and verify it is picked up by the other charm
